packages/core/src/components/tabs2/tabs.tsx

```diff
@@ -0,0 +1,293 @@
+        const focusedIndex = enabledTabElements.indexOf(focusedElement);
+        const direction = this.getKeyCodeDirection(e);
+
+        if (focusedIndex >= 0 && direction !== undefined) {
+            e.preventDefault();
+            const { length } = enabledTabElements;
+            // auto-wrapping at 0 and `length`
+            const nextFocusedIndex = (focusedIndex + direction + length) % length;
+            (enabledTabElements[nextFocusedIndex] as HTMLElement).focus();
+        }
+    }
+
+    private handleKeyPress = (e: React.KeyboardEvent<HTMLDivElement>) => {
+        const targetTabElement = (e.target as HTMLElement).closest(TAB_SELECTOR) as HTMLElement;
+        if (targetTabElement != null && isEventKeyCode(e, Keys.SPACE, Keys.ENTER)) {
+            e.preventDefault();
+            targetTabElement.click();
+        }
+    }
+
+    private handleTabClick = (newTabId: TabId) => {
+        const { selectedTabId } = this.state;
+        if (newTabId !== selectedTabId) {
+            safeInvoke(this.props.onChange, newTabId, selectedTabId);
+            if (this.props.selectedTabId === undefined) {
+                this.setState({ selectedTabId: newTabId });
+            }
+        }
+    }
+
+    private handleTabRef = (tabElement: HTMLDivElement) => { this.tablistElement = tabElement; };
+
+    /**
+     * Calculate the new height, width, and position of the tab indicator.
+     * Store the CSS values so the transition animation can start.
+     */
+    private moveSelectionIndicator() {
+        if (this.tablistElement === undefined) { return; }
+
+        const tabIdSelector = `${TAB_SELECTOR}[data-tab-id="${this.state.selectedTabId}"]`;
+        const selectedTabElement = this.tablistElement.query(tabIdSelector) as HTMLElement;
+
+        let indicatorWrapperStyle: React.CSSProperties = { display: "none" };
+        if (selectedTabElement != null) {
+            const { clientHeight, clientWidth, offsetLeft, offsetTop } = selectedTabElement;
+            indicatorWrapperStyle = {
+                height: clientHeight,
+                transform: `translateX(${Math.floor(offsetLeft)}px) translateY(${Math.floor(offsetTop)}px)`,
+                width: clientWidth,
+            };
+        }
+        this.setState({ indicatorWrapperStyle });
+    }
+
+    private renderTabPanel = (tab: TabElement) => {
+        const { className, panel, id } = tab.props;
+        if (panel === undefined) {
+            return undefined;
+        }
+        return (
+            <div
+                aria-labelledby={generateTabTitleId(this.props.id, id)}
+                aria-hidden={id !== this.state.selectedTabId}
+                className={classNames(Classes.TAB_PANEL, className)}
+                id={generateTabPanelId(this.props.id, id)}
+                key={id}
+                role="tabpanel"
+            >
+                {panel}
+            </div>
+        );
+    }
+
+    private renderTabTitle = (tab: TabElement) => {
```

Comment:
oops my bad... I guess I'm so used to seeing `this.props` destructured in the other codebases I work in. carry on